### ISSUE TITLE
The refine() function for a boolean is causing compilation error

### ISSUE THREAD

**Issue by Alediran** [colinhacks/zod#620]
I am coding a project with the following libraries:

    "@hookform/resolvers": "^2.7.1",
    "@reduxjs/toolkit": "^1.6.1",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^9.5.0",
    "@testing-library/user-event": "^7.2.1",
    "@types/jest": "^24.9.1",
    "@types/node": "^12.20.16",
    "@types/react": "^16.14.11",
    "@types/react-dom": "^16.9.14",
    "@types/react-redux": "^7.1.18",
    "axios": "^0.21.1",
    "node-sass": "^6.0.1",
    "npm": "^7.20.0",
    "primeflex": "^2.0.0",
    "primeicons": "^4.1.0",
    "primereact": "^6.5.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-localization": "^1.0.17",
    "react-redux": "^7.2.4",
    "react-router-dom": "^5.2.0",
    "react-scripts": "^4.0.3",
    "react-transition-group": "^4.4.2",
    "typescript": "~4.1.5",
    "zod": "^3.8.1"

I defined a UserCreateDto to handle and validate the most basic data for a Create User form. I also created a customized errorMap.

user-create.dto.tsx:

```
import { z } from 'zod';
import Localize from '../../components/localize';
import errorMap from '../error.map';

z.setErrorMap(errorMap);

export const validationSchema = z
	.object({
		User: z.string().min(8, { message: Localize['Validation:UserName'] }),
		Email: z.string().email({ message: Localize['Validation:InvalidEmail'] }),
		Password: z.string().min(1, { message: Localize['Validation:Required'] }),
		confirm: z.string().min(1, { message: Localize['Validation:Required'] }),
		Birthday: z
			.date()
			.optional()
			.refine((data) => data !== undefined, {
				message: Localize['Validation:UndefinedDate'],
			}),
		Accept: z.boolean().refine((data) => data === true, { message: '' }),
	})
	.refine((data) => data.Password === data.confirm, {
		message: Localize['Validation:PassswordsNotMatching'],
		path: ['confirm'],
	});

type UserCreateDto = z.infer<typeof validationSchema>;

export default UserCreateDto;
```

error.map.tsx

```
import { z } from 'zod';
import Localize from '../components/localize';

const errorMap: z.ZodErrorMap = (issue, ctx) => {
	if (issue.code === z.ZodIssueCode.invalid_type) {
		if (issue.expected === 'string') {
			return { message: Localize['Validation:Required'] };
		}
		if (issue.expected === 'date') {
			return { message: Localize['Validation:Required'] };
		}
	}
	if (issue.code === z.ZodIssueCode.custom) {
		return { message: `less-than-${(issue.params || {}).minimum}` };
	}
	return { message: ctx.defaultError };
};

export default errorMap;
```

The error I get when I do either npm start or npm run build is this:

Failed to compile.

undefined
fork-ts-checker-webpack-plugin error in undefined(undefined,undefined):
Maximum call stack size exceeded  TSINTERNAL

The error disappears when I remove .refine((data) => data === true, { message: '' }) from the Accept boolean property in the validationSchema. I can also add it back to the validation after npm start completes the startup.

**Comment by colinhacks** [comment]
Sorry, not able to reproduce on my machine. Perhaps try putting together a reproduction on [codesandbox.io](codesandbox.io).

**Comment by colinhacks** [comment]
I was able to reproduce this finally. This should be resolved in Zod 3.9.
